Call conversion functions directly

martins-mozeiko opened the pull request:
This makes generated convert to and from function to call C functions directly instead of going through Python capsule API - that currently requires importing module, which can lead to a lot of imports for nested types in array. Now conversion functions are placed in separate libraries that link to each other. Python module libraries link to these new libraries.

See #9 for more details.

Sorry, I cannot provide any specific performance numbers right now, I can try to do that a bit later.

I have tested this change on Ubuntu 16 and Windows.

Connects to #9

mikaelarguedas commented:
A run of CI on all platforms. It looks like the libraries of the dependencies cannot be found on MacOS.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5096)](http://ci.ros2.org/job/ci_linux/5096/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1880)](http://ci.ros2.org/job/ci_linux-aarch64/1880/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4226)](http://ci.ros2.org/job/ci_osx/4226/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5077)](http://ci.ros2.org/job/ci_windows/5077/)

mikaelarguedas commented on rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake:
```diff
@@ -209,6 +219,11 @@ foreach(_typesupport_impl ${_typesupport_impls})
+  add_dependencies(
+    ${_target_name_lib}
+    ${rosidl_generate_interfaces_TARGET}${_target_suffix}
+    ${rosidl_generate_interfaces_TARGET}__rosidl_typesupport_c
```
Along the same lines, if we don't need to depend on specific typesupport, we may also be able to generate a single `_python_` library instead of 1 per typesupport

mikaelarguedas replied:
Ok I gave that a shot at https://github.com/ros2/rosidl_python/commit/7fbb83234e56a1660ac62fef6dd1ad556b847795 and it seems to work fine with a single python library per package

dirk-thomas commented:
@martins-mozeiko Is there any update in this?

martins-mozeiko commented:
I've added changes requested - now it uses dylib extension on macOS (I have not Mac available currently, so I cannot test it), and it uses visibility macros from `visibility_control.h` header instead of defining its own macros. 

I've also integrated changes from Mikael and rebased all commits on top of master.

Tried building & running demo_nodes_py talker/listener - it works fine on Windows & Linux.

mikaelarguedas commented:
Thanks @martins-mozeiko for iterating.

> Tried building & running demo_nodes_py talker/listener - it works fine on Windows & Linux.

It looks like this change fails the linter tests on CI, you should be able to test the changes by running `colcon test --packages-select <PACKAGES_TO_TEST>`.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5155)](http://ci.ros2.org/job/ci_linux/5155/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1932)](http://ci.ros2.org/job/ci_linux-aarch64/1932/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4280)](http://ci.ros2.org/job/ci_osx/4280/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5136)](http://ci.ros2.org/job/ci_windows/5136/)

I pushed some fixups:
- spurious line change from my CMake code shuffling: https://github.com/ros2/rosidl_python/pull/10/commits/77009ae53162e2c1ae6c53bae307cff417820351
- remove duplication of visibility_control include: https://github.com/ros2/rosidl_python/pull/10/commits/617e0c67c0a26509e6c0a110aed16bb2d2aa1f76
- remove new blank line in generated files to fix linter failures: https://github.com/ros2/rosidl_python/pull/10/commits/f931dbd5662fc7c3385aa97f93143fa247576148

With fixups (remainung rcl test failure expected and tracked at https://github.com/ros2/rmw_fastrtps/issues/226):
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5156)](http://ci.ros2.org/job/ci_linux/5156/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1933)](http://ci.ros2.org/job/ci_linux-aarch64/1933/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4282)](http://ci.ros2.org/job/ci_osx/4282/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5137)](http://ci.ros2.org/job/ci_windows/5137/)

mikaelarguedas commented:
> Sorry, I cannot provide any specific performance numbers right now, I can try to do that a bit later.

I didnt get a chance to test the performances improvements either. Do you happen to have more data in that matter?

mikaelarguedas commented:
Change looks good to me :+1:

@dirk-thomas do you want to provide a second review on this ? or should I merge as is ?

dirk-thomas commented on rosidl_generator_py/resource/_msg_support.c.em:
```diff
@@ -52,7 +54,22 @@ if not field.type.is_primitive_type() and field.type.is_array:
+@[    if spec.base_type.pkg_name != field.type.pkg_name]@
+ROSIDL_GENERATOR_C_IMPORT
+@[    end if]@
+bool @(field.type.pkg_name)_@(lowercase_field_type)__convert_from_py(PyObject * _pymsg, void * _ros_message);
```
I think this function should use a double underscore between the package name and the field type in order to clearly separate the two from each other (same as between the field type and the fixed name).

Same for `convert_to_py`.

mikaelarguedas replied:
resolved in https://github.com/ros2/rosidl_python/commit/18d2af12adc34c322bf2e8ec8d49e66e62460e19

mikaelarguedas commented:
>  think it would be good to reproduce the performance impact (since that is the whole reason for this patch).

I agree that having numbers regarding the performance improvement would be valuable. However, I think we agree that the new approach is cleaner / preferable to the current one regardless of the exact measured performance gain.

> I think this function should use a double underscore between the package name and the field type in order to clearly separate the two from each other (same as between the field type and the fixed name).

I agree that would be are better naming pattern. Though this PR doesn't change the function names that are already implemented using this scheme. I am fine renaming them in here but would not be opposed in doing it in a follow-up as it's not strictly in the scope of this PR

mikaelarguedas commented:
Renamed the functions in https://github.com/ros2/rosidl_python/pull/10/commits/18d2af12adc34c322bf2e8ec8d49e66e62460e19

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5172)](http://ci.ros2.org/job/ci_linux/5172/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1946)](http://ci.ros2.org/job/ci_linux-aarch64/1946/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4298)](http://ci.ros2.org/job/ci_osx/4298/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5155)](http://ci.ros2.org/job/ci_windows/5155/)

mikaelarguedas commented:
@martins-mozeiko when you get a chance, can you please test the latest state of this PR and provide feedback on the (hopefully) time improvements compared to master?

martins-mozeiko commented:
I've made a following test - publisher sends geometry_msgs/PoseArray messages pretty much as fast as it can. And subscribe counts how long does it take to receive 100 messages.

Here is publisher.py code:
```
#!/usr/bin/env python3

import rclpy, rclpy.node
from geometry_msgs.msg import Pose, PoseArray

count = 0

def cb():
  global pub
  global count
  global msg

  pub.publish(msg)

  count += 1
  print("Sent", count)

msg = PoseArray()
msg.header.stamp.sec = 100
msg.header.stamp.nanosec = 200
msg.header.frame_id = "frame"
for i in range(1024):
  p = Pose()
  p.position.x = 1.0
  p.position.y = 2.0
  p.position.z = 3.0
  p.orientation.x = 4.0
  p.orientation.y = 5.0
  p.orientation.z = 6.0
  p.orientation.w = 7.0
  msg.poses.append(p)

rclpy.init()
n = rclpy.node.Node("publisher")
pub = n.create_publisher(PoseArray, "/test")
timer = n.create_timer(0.001, cb)

rclpy.spin(n)
```

And here is subscriber.py code:
```
#!/usr/bin/env python3

import rclpy, rclpy.node
from time import time
from geometry_msgs.msg import PoseArray

count = 0

def cb(msg):
  global count
  count += 1
  if count == 100:
    global start
    print(time() - start)
    exit()

rclpy.init()
n = rclpy.node.Node("listener")
sub = n.create_subscription(PoseArray, "/test", cb)

start = time()
rclpy.spin(n)
```

Running this on Raspberry Pi 3 with ROS2 built from master gives me **~63 seconds**.

After applying this pull request the time decreases to **~37 seconds**.

mikaelarguedas commented:
Tried a similar thing on amd64 with a few changes:
- setting the start time when the first received message so that it's easier to reproduce
- printing to console only once every 100th
- stop after receiving a thousand message

Without this patch
```
$ PYTHONOPTIMIZE=0 python3 receiver.py 
16.82235884666443
```
With this patch:
```
$ PYTHONOPTIMIZE=0 python3 receiver.py 
14.596002578735352
```

<details><summary>Code</summary>

sender.py:
```
import rclpy, rclpy.node
from geometry_msgs.msg import Pose, PoseArray

count = 0

def cb():
    global pub
    global count
    global msg

    pub.publish(msg)

    count += 1
    if count % 100 == 0:
        print("Sent", count)

msg = PoseArray()
msg.header.stamp.sec = 100
msg.header.stamp.nanosec = 200
msg.header.frame_id = "frame"
for i in range(1024):
    p = Pose()
    p.position.x = 1.0
    p.position.y = 2.0
    p.position.z = 3.0
    p.orientation.x = 4.0
    p.orientation.y = 5.0
    p.orientation.z = 6.0
    p.orientation.w = 7.0
    msg.poses.append(p)

rclpy.init()
n = rclpy.create_node("publisher")
pub = n.create_publisher(PoseArray, "/test")
timer = n.create_timer(0.001, cb)

rclpy.spin(n)
```

receiver.py:
```
import rclpy
from time import time
from geometry_msgs.msg import PoseArray

count = 0
start = None


def cb(msg):
    global count
    global start
    if start is None:
        start = time()

    count += 1
    if count == 1000:
        global start
        print(time() - start)
        exit()


rclpy.init()
n = rclpy.create_node('listener')
sub = n.create_subscription(PoseArray, '/test', cb)

rclpy.spin(n)

```

</details>

Not as drastic as what is observed on the RaspberryPi but there is definitely an improvement

mikaelarguedas commented:
Thanks @martins-mozeiko for the patch :+1:
